datafusion/sql/src/unparser/plan.rs

```diff
@@ -129,14 +129,88 @@ impl Unparser<'_> {
                 Ok(())
             }
             LogicalPlan::Projection(p) => {
-                let items = p
-                    .expr
-                    .iter()
-                    .map(|e| self.select_item_to_sql(e))
-                    .collect::<Result<Vec<_>>>()?;
-                select.projection(items);
-
-                self.select_to_sql_recursively(p.input.as_ref(), query, select, relation)
+                // A second projection implies a derived tablefactor
+                if !select.already_projected() {
```

Comment:
This block should probably be refactored into multiple helper functions to reduce cognitive load here...